src/bindings/c/include/openvino/c/ov_property.h

```diff
@@ -13,7 +13,7 @@
 
 #include "openvino/c/ov_common.h"
 
-typedef struct ov_property ov_property_t;
+typedef struct ov_properties ov_properties_t;
```

Comment:
Can on C level this structure be like `char*, ov_property_value_t`? And we will use array of these properties in application code. In this case we don't need to have methods like: `ov_properties_add` `ov_properties_size` and etc.
@riverlijunjie What do you think?

Reply:
If so, how to handle the case that we need add indeterminate number items into properties? maybe need reallocate the buffer for new items or add a "next" ptr to point to the next item and allocate item one by one? Both of them will bring more additional method than current implement.

Reply:
But I think in your application you always know how much properties you want to set :)
In this case you can just allocate an array of properties with the right size.

Reply:
Maybe we cannot suppose that user always know accurate property item number in some special scenario? For example, in some application, it will collect some information to determinate add which property item, it will do it one by one, and it will wired that app first scan all information to get properties item number and then allocate its memory to do the 2nd scan to fill the properties.
Anyhow, I think property interface is provided to user to fill properties, and it will be flexible if it is dynamic buffer.

Reply:
The problem is that C doesn't provide `vector` and other containers, if user use C, I think he is ready to work with raw arrays.
In case if user would like to add some additional properties in the process of application execution, he can just call one more time get_property method for new set of properties.

Reply:
I don't think it is a good reason, which still bring trouble to customer. 
1). we have to add more logic to manage key string's memory, which is easy lead to memory leak.
2). property value is defined as ov_any { *ptr; size;  type }, which need memory management, if string key is applied , it will make more  complicated. maybe user will say "oh, my God, how hard to use such interface." 
3). we have to add logic to compare string key one by one, If i'm right we cannot use switch for std::string, and have to use a long if..else to do it. 

@ilya-lavrenov @wangleis could you give some suggestion? It seems that Ilya.C and I cannot convince each other.

Reply:
If user develop own application on C language, he should know how memory management works on C level. We shouldn't provide C++ API inside C.

С and C++ development approaches are difference. let's stick to language guidelines instead of introducing unclear hidden structures just because we can.

Moreover let's think first of all about the use cases. we have spend already 6+ weeks on code review on these PRs, with different attempts to explain and proof where need to extend API, when we have use cases and we need to have more flexible API, or decrease complexity of hidden API in cases which we don't have any real cases.

In this particular case I don't believe that somebody will generate dynamic list of properties and after that load it into core.
If user need to add some properties, he can multiple time call `set_property()` method.

Reply:
We cannot suppose the application behavior, the mature apis should provide flexible usage support. Anyhow, I don't think you are wrong or I'm wrong. Maybe we need make a final decision which one will be chosen:

proposal 1:  current implement, using an opaque of ov::AnyMap to represent C property list and add new items dynamically.
proposal 2: define property list as a C structure {key, value} list, and pre-allocate memory for the all items in the property list, in which  value is ov_type_t = {void *ptr, size, type} also need allocate additional buffer. 

@ilya-lavrenov  could you vote it?

Reply:
@riverlijunjie The problem is that in the first solution you want to sell C++ under the C API.

Moreover in first solution you also need to allocate the memory for value ov_any_t:
```
OPENVINO_C_API(ov_status_e)
ov_properties_add(ov_properties_t* property, ov_property_key_e key, ov_any_t* value);
```

First and second approaches have only 1 big difference:

- First solution hides implementation and have "magic" method which allow to add new element. The second proposal is more aligned with C guidelines: you allocate the memory for the array of properties with fixed size and after that work with it. In this solution we provide C convenient way to work with properties.

Moreover in other PRs for example for dimensions and tensors we already moved to the same approach.

Here in order to understand what solution is better let's start from user perspective, and at the current moment we don't have use cases on customer side where we need to have dynamic set of properties (you can take a look to our samples and demos).
And if in the future we will have such cases where customer in the process of execution will add new properties it is not a big deal to have several calls of get/set_property.
Also C way solution allows to iterate over all properties using standard syntax `properties[n]`.
Current solution doesn't provide such solutions.

So my main point is C is not C++, it is two different languages, if user wants to develop on C++, we have such API. If user wants to develop C application, let's provide such API, which respects languages guidelines and doesn't try to simulate behaviour of other languages.

So from my point of view on such discussions we already waste tons of time and each hour increases the risk from these changes, because honestly we don't have a time for a good validation (honestly we don't have a time for ANY validation), but we need to provide a good quality of product...
I think customer would like to use API which doesn't have bugs, memory leaks or something like that.

Reply:
If we want to respect language guideline, we should also implement pure C language for each C++ class/object, such as ov::Tensor and ov::Core, ov::Model should be also C implement other than encapsulated to be C interface. But it is impossible because it will bring tons of complicated transformation.
Actually, we call it as a C wrapper but not C implement, a simple/effective/friendly/easy-use interface definition is the first requirement.
ok, let's follow the proposal 2.

Reply:
If we are talking about `get` methods, we can use OpenCL approach https://registry.khronos.org/OpenCL/sdk/1.1/docs/man/xhtml/clSetKernelArg.html, but OpenCL uses `[cl_uint](https://registry.khronos.org/OpenCL/sdk/1.1/docs/man/xhtml/scalarDataTypes.html) arg_index` to define argument, we need to use either char* (@ilyachur approach) or enum (@riverlijunjie approach).

If we want to set, I suppose we have 2 options:
- have an array of properties
- have a dedicated object

Reply:
Looks like we can have array of properties instead of magic property. Could you please in pseudo code how 2 approaches will look like ? E.g. I want to set 2 properties to ov_core_compile_model

Reply:
Option 1: property array
    ov_properties_t properties;
    ov_properties_create(&properties, 2); // allocate buffer for the array

    const char* key_1 = ov_property_key_hint_performance_mode;
    ov_performance_mode_e mode = ov_performance_mode_e::THROUGHPUT;
    ov_any_t value_1 = {(void*)&mode, 1, ov_any_type_e::ENUM};
    properties.list[0].key = key_1;
    properties.list[0].value = value_1;

    const char* key_2 = ov_property_key_cache_dir;
    const char cache_dir[] = "./cache_dir";
    ov_any_t value_2 = {(void*)cache_dir, sizeof(cache_dir), ov_any_type_e::CHAR};
    properties.list[1].key = key_2;
    properties.list[1].value = value_2;

    ov_core_set_property(core, device_name.c_str(), &properties);
    ov_properties_free(&properties);




Option 2: a dedicated object:

    ov_properties_t* properties = nullptr;
    ov_properties_create(&properties);   // Create opaque handle, but not allocate actual buffer now
        
    const char* key_1 = ov_property_key_hint_performance_mode;
    ov_performance_mode_e mode = ov_performance_mode_e::THROUGHPUT;
    ov_any_t value_1 = {(void*)&mode, 1, ov_any_type_e::ENUM};
    ov_properties_add(property, key_1, value_1);

    const char* key_2 = ov_property_key_cache_dir;
    const char cache_dir[] = "./cache_dir";
    ov_any_t value_2 = {(void*)cache_dir, sizeof(cache_dir), ov_any_type_e::CHAR};
    ov_properties_add(property, key_2, value_2);

     ov_core_set_property(core, device_name.c_str(), &properties);
     ov_properties_free(&properties);

Reply:
Option 2 is previous implement.
Now I have changed to Option 1.